### ISSUE TITLE
Get-DbaCmObject: Possible error handling issue / never-ending loop

### ISSUE THREAD

**Issue by SQLGB** [sqlcollaborative/dbatools#5998]
### Environmental information

```#### PowerShell version:
Name                           Value                                                                                                                                                        
----                           -----                                                                                                                                                        
PSVersion                      5.1.14409.1018                                                                                                                                               
PSEdition                      Desktop                                                                                                                                                      
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                      
BuildVersion                   10.0.14409.1018                                                                                                                                              
CLRVersion                     4.0.30319.42000                                                                                                                                              
WSManStackVersion              3.0                                                                                                                                                          
PSRemotingProtocolVersion      2.3                                                                                                                                                          
SerializationVersion           1.1.0.1                                                                                                                                                      

#### dbatools Module version:
Name    : dbatools
Path    : C:\Program Files\WindowsPowerShell\Modules\dbatools\1.0.36\dbatools.psd1
Version : 1.0.36
```

```powershell
PS C:\Windows\system32> Get-Culture

LCID             Name             DisplayName                                                                                                                                               
----             ----             -----------                                                                                                                                               
1033             en-US            English (United States)                                                                                                                                   

```

#### SQL Server: 

```sql
Microsoft SQL Server 2016 (SP1-CU15-GDR) (KB4505221) - 13.0.4604.0 (X64)   Jun 15 2019 07:56:34   Copyright (c) Microsoft Corporation  Enterprise Edition (64-bit) on Windows Server 2016 Datacenter 10.0 <X64> (Build 14393: ) (Hypervisor) 
```

```sql
us_english
```

#### Host used

- [x] powershell.exe
- [x] ISE
- [ ] VS Code
- [ ] Other (please specify)

If anything other than powershell.exe was used, please confirm that you can duplicate the issue with powershell.exe

- [x] Still buggy in powershell.exe

#### Errors Received

```
PSMessageDetails      : 
Exception             : System.Management.Automation.RuntimeException: An error occurred while enumerating through a collection: The WS-Management 
                        service cannot process the request. This service has exceeded its maximum of 2147483647 concurrent operations allowed for 
                        plugin "WMI Provider". Retry the request later or raise the Maximum Shells per Plugin quota. . ---> 
                        Microsoft.Management.Infrastructure.CimException: The WS-Management service cannot process the request. This service has 
                        exceeded its maximum of 2147483647 concurrent operations allowed for plugin "WMI Provider". Retry the request later or raise 
                        the Maximum Shells per Plugin quota. 
                           at Microsoft.Management.Infrastructure.Internal.Operations.CimSyncEnumeratorBase`1.MoveNext()
                           at System.Management.Automation.EnumerableOps.MoveNext(ExecutionContext context, IEnumerator enumerator)
                           --- End of inner exception stack trace ---
                           at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception 
                        exception)
                           at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
                           at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
                           at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
TargetObject          : Microsoft.Management.Infrastructure.Internal.Operations.CimSyncInstanceEnumerator
CategoryInfo          : InvalidOperation: (Microsoft.Manag...tanceEnumerator:CimSyncInstanceEnumerator) [], RuntimeException
FullyQualifiedErrorId : BadEnumeration
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Get-DbaCmObject<Process>, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.0.36\allcommands.ps1: line 20027
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}

```

#### Steps to Reproduce

```
Get-DbaCmObject -Computername "MyServer" -ClassName Win32_OperatingSystem -debug
```

#### Expected Behavior

After occurrence of the following, I would have expected the command to abandon Cim over WinRM connection attempts and move on to Cim over DCOM (or something else) but, it continues until the command is cancelled.  Given the content of the exception reported above in the errors section, it is possible that this is being influenced by a server with WMI in a problem state.  If possible, it would be nice is dbatools could handle this state.
```
DEBUG: 20025 | [14:25:07][Get-DbaCmObject] [MyServer] Accessing computer using Cim over WinRM
DEBUG: 20036 | [14:25:07][Get-DbaCmObject] [MyServer] Accessing computer using Cim over WinRM - Failed
```

#### Actual Behavior

Command continues to output error / status messages as shown above until it is cancelled.

**Comment by mjbarlow1985** [comment]
This sounds very similar to an issue i have running  Get-DbaWsfcCluster. This in turn runs Get-DbaCmObject  -Namespace root\MSCluster -ClassName MSCluster_Cluster. It just hangs. Whilst i suspect there may be an issue on the machine. I cant handle it because it just hangs. My job has unintentionally ran for days over the weekend, it would just not stop.  Ill keep my eye on this....

**Comment by SQLGB** [comment]
@mjbarlow1985 Nice to know it's not just me.

I scanned the "allcommands.ps1" file and found references to Get-DbaCmObject in 35 commands / functions (some of which are "external helper" functions); potentially all would be impacted by the reported issue:
```
Get-DbaClientProtocol
Get-DbaComputerSystem
Get-DbaCpuUsage
Get-DbaDiskSpace
Get-DbaFilestream
Get-DbaInstanceInstallDate
Get-DbaInstanceProtocol
Get-DbaOperatingSystem
Get-DbaPageFileSetting
Get-DbaPowerPlan
Get-DbaService
Get-DbaWsfcAvailableDisk
Get-DbaWsfcCluster
Get-DbaWsfcNetwork
Get-DbaWsfcNetworkInterface
Get-DbaWsfcNode
Get-DbaWsfcResource
Get-DbaWsfcResourceType
Get-DbaWsfcRole
Get-DbaWsfcSharedVolume
Install-DbaInstance
Measure-DbaDiskSpaceRequirement
New-DbaComputerCertificate
New-DbatoolsSupportPackage
Reset-DbaAdmin
Resolve-DbaNetworkName
Set-DbaPowerPlan
Test-DbaPowerPlan
Get-DbaReportingService
Get-InternalService
Get-SqlInstanceUpdate
Set-FileSystemSetting
Set-ServiceStartMode
Test-PendingReboot
Update-ServiceStatus
```

**Comment by wsmelton** [comment]
I'm not aware of any method in CIM to track that a server is allowing connection but not responding in a proper way. 

@FriedrichWeinmann may know how we can better handle this situation.

**Comment by FriedrichWeinmann** [comment]
Well ... it sure is a bug in Get-DbaCmObject and I'll look into it right away.
@SQLGB: Thank you for the very thorough description, much appreciated and it will help :)

Just for reference: The 15 minutes timeout is something slightly different:
By default, when all connections fail, it marks the server as unreachable. If you try to connect to it again right away (as many of our more complex commands may do as part of their workflow), it will not even try to connect - it just fails right away. Only after the timeout has expired, it will try again to connect to the host (and if that fails again, the timeout starts anew).

**Comment by FriedrichWeinmann** [comment]
Found the issue. And the culprit was ... me, alas :(
When I was dealing with the last error handling problem I did a full refactor of the error classification and for hell's sake forgot to include a default action.
So whenever it would have an unaccounted for error - I accounted for every single error code in the CIM system, so it'd take some issue outside of it, as this one was - it would have just that problem.

PR #6012 should fix that

**Comment by SQLGB** [comment]
@FriedrichWeinmann Thanks for the status update and your quick response to the issue!

**Comment by wsmelton** [comment]
Reopening after talking with SQLGB via another issue #6003 (think these are related in some fashion based on the output.

**Comment by SQLGB** [comment]
@wsmelton I think you can re-close 5998; release 1.0.39 appears to have resolved this issue based on my initial tests.  The output below is an example of Get-DbaCmObject now recognizing a failure with Cim over WinRM and then moving on to the next option:

```
PS C:\Windows\system32> Get-DbaCmObject -ComputerName "MyServer" -ClassName Win32_OperatingSystem -debug
DEBUG: 20075 | [09:16:21][Get-DbaCmObject] Configuration loaded | Cache disabled: False
DEBUG: 20088 | [09:16:21][Get-DbaCmObject] [MyServer] Retrieving Management Information
DEBUG: 20128 | [09:16:21][Get-DbaCmObject] [MyServer] Accessing computer using Cim over WinRM
DEBUG: 20139 | [09:16:21][Get-DbaCmObject] [MyServer] Accessing computer using Cim over WinRM - Failed
DEBUG: 20240 | [09:16:21][Get-DbaCmObject] [MyServer] Accessing computer using Cim over DCOM

DEBUG: 20245 | [09:16:21][Get-DbaCmObject] [MyServer] Accessing computer using Cim over DCOM - Success
SystemDirectory     Organization BuildNumber RegisteredUser SerialNumber            Version   
---------------     ------------ ----------- -------------- ------------            -------   
C:\Windows\system32              14393       Windows User   00376-40000-00000-AA947 10.0.14393

```

**Comment by mjbarlow1985** [comment]
To confirm, the issue i described in my previous comment is also resolved using 1.0.39. Many thanks.

**Comment by potatoqualitee** [comment]
Sweet! Thanks for the update!
